x/ibc-hooks/contracts/crosschain-swaps/src/contract.rs

```diff
@@ -0,0 +1,115 @@
+
+// version info for migration info
+const CONTRACT_NAME: &str = "crates.io:crosschain-swaps";
+const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+/// Handling contract instantiation
+#[cfg_attr(not(feature = "imported"), entry_point)]
+pub fn instantiate(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    msg: InstantiateMsg,
+) -> Result<Response, ContractError> {
+    set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+
+    // validate contract addresses and save to config
+    let swap_contract = deps.api.addr_validate(&msg.swap_contract)?;
+    let state = Config {
+        swap_contract,
+        track_ibc_callbacks: msg.track_ibc_sends.unwrap_or(false),
+    };
+    CONFIG.save(deps.storage, &state)?;
+    for (prefix, channel) in msg.channels.into_iter() {
+        CHANNEL_MAP.save(deps.storage, &prefix, &channel)?;
+    }
+
+    Ok(Response::new()
+        .add_attribute("method", "instantiate")
+        .add_attribute("owner", info.sender))
+}
+
+#[cfg_attr(not(feature = "imported"), entry_point)]
+pub fn migrate(_deps: DepsMut, _env: Env, msg: MigrateMsg) -> Result<Response, ContractError> {
+    match msg {}
+}
+
+/// Handling contract execution
+#[cfg_attr(not(feature = "imported"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    match msg {
+        ExecuteMsg::OsmosisSwap {
+            input_coin,
+            output_denom,
+            receiver,
+            slippage,
+            next_memo,
+            failed_delivery,
+        } => execute::swap_and_forward(
+            deps,
+            env.block.time,
+            env.contract.address,
+            input_coin,
+            output_denom,
+            slippage,
+            receiver,
+            next_memo,
+            failed_delivery,
+        ),
+        ExecuteMsg::Recover {} => execute::recover(deps, info.sender),
+    }
+}
+
+/// Handling contract queries
+#[cfg_attr(not(feature = "imported"), entry_point)]
+pub fn query(deps: Deps, _env: Env, msg: QueryMsg) -> StdResult<Binary> {
+    match msg {
+        QueryMsg::Recoverable { addr } => to_binary(
+            &RECOVERY_STATES
+                .may_load(deps.storage, &addr)?
+                .or(Some(vec![])),
+        ),
+    }
+}
+
+#[cfg_attr(not(feature = "imported"), entry_point)]
+pub fn sudo(deps: DepsMut, _env: Env, msg: SudoMsg) -> Result<Response, ContractError> {
+    match msg {
+        SudoMsg::ReceivePacket {} => unimplemented!(),
+        SudoMsg::ReceiveAck {
+            channel,
+            sequence,
+            ack,
+            success,
+        } => sudo::receive_ack(deps, channel, sequence, ack, success),
+        SudoMsg::ReceiveTimeout {} => unimplemented!(),
+    }
+}
+
+#[cfg_attr(not(feature = "imported"), entry_point)]
+pub fn reply(deps: DepsMut, _env: Env, reply: Reply) -> Result<Response, ContractError> {
+    deps.api
```

Comment:
Wanna keep `deps.api.debug` on? I guess there is no problem since on mainnet it's gonna be disabled anyway, but there are commented ones elsewhere, so I think, either remove them all or uncomment other ones

Reply:
Same as with the comments above: I have mixed feelings about that. 

I think it makes sense to keep some on and some commented out for future debugging. I also don't think that call has a performance impact on prod

Reply:
I think this is my ignorance of cosmwasm -- who is this entrypoint for? Synchronous calls from other contracts?

Reply:
TY for the detailed explanation! Makes sense